### PR TITLE
Smarti inline highlighting

### PR DESCRIPTION
**Pull request by mrsimpson**
This brings highlighting of found terms within the text to Assistify
smarti highlighting

# Known issues
- [x] Once an obsolete search term was highlighted and is clicked, we need to re-add this term.

### CODE REVIEW

**Comment by mrsimpson** [comment]
Finally, a last comment before going to sleep: There was a modification necessary at the `room.js` file.
Though the message itself is the one emitting the event (`click .recognized-term`), the event cannot be handled by the message: The message template is rendered non-reactive. Therefore, it also emits no events, but the event "bubble" to the parent template - which is the room.
While there are extension mechanisms on the message rendering side, I was missing those on the message interaction part.
The fact, that mentions are handled equally, reduces hope I have overseen something...

File: packages/assistify-ai/client/messageRenderer.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * This function is used as a callback which is executed while rendering the message.
+ * @see /packages/rocketchat-ui-message/client/renderMessageBody.js
+ * @param {*} message the message object including all properties
+ * @augments message.html - the rendered message body
+ */
+const highlightRecognizedTerms = function(message) {
+	if (RocketChat.settings.get('Assistify_AI_Smarti_Inline_Highlighting_Enabled')) {
+		let { html } = message;
+		if (message.recognizedTerms) {
+			message.recognizedTerms.forEach((term) => html = html.replace(new RegExp(`(^|\\b|[\\s\\n\\r\\t.,،'\\\"\\+!?:-])(${ s.escapeRegExp(term) })($|\\b|[\\s\\n\\r\\t.,،'\\\"\\+!?:-])(?![^<]*>|[^<>]*<\\/)`, 'gmi'), '$1<span class="recognized-term">$2</span>$3'));
```
Comment (jschirrmacher) [inline packages/assistify-ai/client/messageRenderer.js]: Isn't at least the non-trivial RegExp worth to spend more than one line?

File: packages/assistify-ai/server/lib/SmartiAdapter.js [hunk 2]
```diff
@@ -129,6 +129,11 @@ export class SmartiAdapter {
 			// we at least cannot consider the sync successful
 			Meteor.defer(() => SmartiAdapter._markRoomAsUnsynced(message.rid));
 		}
+
+		// TODO: Setting
```
Comment (jschirrmacher) [inline packages/assistify-ai/server/lib/SmartiAdapter.js]: I don't understand what is to be done here. A setting exists at least.

File: packages/assistify-ai/server/lib/SmartiAdapter.js [hunk 3]
```diff
@@ -238,6 +243,33 @@ export class SmartiAdapter {
 		SystemLogger.debug('Smarti - retieved analysis result =', JSON.stringify(analysisResult, null, 2));
 		SystemLogger.debug(`Smarti - analysis complete -> update cache and notify room [ ${ roomId } ] for conversation [ ${ conversationId } ]`);
 		RocketChat.Notifications.notifyRoom(roomId, 'newConversationResult', analysisResult);
+
+		// update the affected message with the tokens extracted. This will trigger a re-rendering of the message
+		SmartiAdapter.updateTokensInMessages(roomId, analysisResult);
+	}
+
+	static updateTokensInMessages(roomId, analysisResult) {
+		const allTerms = analysisResult.tokens.reduce((terms, token) => terms.add(token.value), new Set());
```
Comment (jschirrmacher) [inline packages/assistify-ai/server/lib/SmartiAdapter.js]: Couldn't we just use `analysisResult.tokens` here? It seems to be an array, and `forEach` which is used on `allTerms` in line 261 also works on those as well.

File: packages/rocketchat-i18n/i18n/assistifyAI.de.i18n.json [hunk 4]
```diff
@@ -11,6 +11,8 @@
     "Assistify_AI_Smarti_Domain_Description": "Trage hier den Smarti Client-Name ein, der zu dieser Rocket.Chat Instanz gehört.",
     "Assistify_AI_Smarti_Auth_Token": "Smarti Client-Token",
     "Assistify_AI_Smarti_Auth_Token_Description": "Trage hier den Smarti Client-Token ein, damit sich Rocket.Chat gegenüber Smarti authentifizieren kann.",
+    "Assistify_AI_Smarti_Inline_Highlighting_Enabled": "Hervorheben identifizierter Termini",
```
Comment (jschirrmacher) [inline packages/rocketchat-i18n/i18n/assistifyAI.de.i18n.json]: We should use `Terme` here like in the message below.  [Duden](https://www.duden.de/rechtschreibung/Term)

File: packages/assistify-ai/server/lib/SmartiAdapter.js (hunk 2 above)
Comment (mrsimpson) [reply packages/assistify-ai/server/lib/SmartiAdapter.js]: forgot to delete the comment after adding the setting ;)

File: packages/rocketchat-ui/client/views/app/room.js [hunk 5]
```diff
@@ -721,6 +721,52 @@ Template.room.events({
...
+	'click .recognized-term'(e) {
+		if (!Meteor.userId()) {
+			return;
+		}
+		const term = $(e.currentTarget).get(0).innerText;
+
+		Meteor.setTimeout(() => {
+			if ($('.messages-container-wrapper .contextual-bar').length === 0) { // the tab bar is closed
+				// open the tab
+				$('div:not(.active) .rc-header .rc-room-actions .rc-room-actions__action[data-id="assistify-ai"] > button').click();
```
Comment (jschirrmacher) [inline packages/rocketchat-ui/client/views/app/room.js]: Could there be more than one `.rc-room-actions__action[data-id="assistify-ai"] > button`? Else I would simplify the expression to contain only this part. Or maybe remove at least `.rc-header .rc-room-actions` if you want to make sure that the button only gets clicked if it is not in an `.active` div.

File: packages/assistify-ai/server/lib/SmartiAdapter.js (hunk 3 above)
Comment (mrsimpson) [reply packages/assistify-ai/server/lib/SmartiAdapter.js]: I wanted to have a `Set` in order to ensure removed duplicates and check value with `has`

File: packages/rocketchat-ui/client/views/app/room.js (hunk 5 above)
Comment (mrsimpson) [reply packages/rocketchat-ui/client/views/app/room.js]: made it more specific for performance reasons

**Pull request merged**